packages/slice-machine/src/hooks/useLocalStorageItem.ts

```diff
@@ -0,0 +1,62 @@
+import { Dispatch, SetStateAction, useMemo, useSyncExternalStore } from "react";
+
+export type UseLocalStorageItemInfo = {
+  /** Whether the item is unset in localStorage and using the defaultValue as a fallback. */
+  isUnset: boolean;
+};
+
+export type UseLocalStorageReturnType<T> = [
+  T,
+  Dispatch<SetStateAction<T>>,
+  UseLocalStorageItemInfo,
+];
+
+const SLICE_MACHINE_STORAGE_PREFIX = "slice-machine";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const isFunction = (value: any): value is (...args: any[]) => any => {
+  return typeof value === "function";
+};
+
+export function useLocalStorageItem<T>(
+  key: string,
+): UseLocalStorageReturnType<T | undefined>;
+export function useLocalStorageItem<T>(
+  key: string,
+  defaultValue: T,
+): UseLocalStorageReturnType<T>;
+export function useLocalStorageItem<T>(
+  key: string,
+  defaultValue?: T | undefined,
+): UseLocalStorageReturnType<T | undefined> {
+  const storageKey = `${SLICE_MACHINE_STORAGE_PREFIX}_${key}`;
+  const getSnapshot = () => localStorage.getItem(storageKey);
+
+  const subscribe = (listener: () => void) => {
+    window.addEventListener("storage", listener);
+    return () => void window.removeEventListener("storage", listener);
+  };
+
+  const serializedItem = useSyncExternalStore(subscribe, getSnapshot);
+  const item = useMemo(() => {
+    try {
+      if (serializedItem != null) return JSON.parse(serializedItem) as T;
+      return defaultValue;
+    } catch (error) {
+      console.warn(`Error reading localStorage key “${storageKey}”:`, error);
+      return defaultValue;
+    }
+  }, [serializedItem, defaultValue, storageKey]);
+
+  const setItem = (value: SetStateAction<T | undefined>) => {
+    try {
+      const newValue = JSON.stringify(isFunction(value) ? value(item) : value);
+      localStorage.setItem(storageKey, newValue);
+      window.dispatchEvent(new StorageEvent("storage", { key, newValue }));
+    } catch (error) {
+      console.warn(`Error setting localStorage key “${storageKey}”:`, error);
+    }
+  };
+
+  return [item, setItem, { isUnset: serializedItem == null }];
```

Comment:
So this was mostly to avoid having extra null check around the `status` value. By passing it the initial value, it will always have a value and you can use it freely, when spreading the current value and so on.

Reply:
I also thought about, since this is used to manage a local storage item, having the `defaultValue` be written to the storage if it's unset there (maybe having a prop to tell it to do so), that way we'll always have something to work with.